src/elevenlabs/conversational_ai/conversation.py

```diff
@@ -0,0 +1,210 @@
+        self.callback_agent_response_correction = callback_agent_response_correction
+        self.callback_user_transcript = callback_user_transcript
+        self.callback_latency_measurement = callback_latency_measurement
+
+    def start_session(self):
+        """Starts the conversation session.
+
+        Will run in background thread until `end_session` is called.
+        """
+        ws_url = self._get_signed_url() if self.requires_auth else self._get_wss_url()
+        self._thread = threading.Thread(target=self._run, args=(ws_url,))
+        self._thread.start()
+
+    def end_session(self):
+        """Ends the conversation session."""
+        self.audio_interface.stop()
+        self._should_stop.set()
+
+    def wait_for_session_end(self) -> Optional[str]:
+        """Waits for the conversation session to end.
+
+        You must call `end_session` before calling this method, otherwise it will block.
+
+        Returns the conversation ID, if available.
+        """
+        if not self._thread:
+            raise RuntimeError("Session not started.")
+        self._thread.join()
+        return self._conversation_id
+
+    def _run(self, ws_url: str):
+        with connect(ws_url) as ws:
+
+            def input_callback(audio):
+                ws.send(
+                    json.dumps(
+                        {
+                            "user_audio_chunk": base64.b64encode(audio).decode(),
+                        }
+                    )
+                )
+
+            self.audio_interface.start(input_callback)
+            while not self._should_stop.is_set():
+                try:
+                    message = json.loads(ws.recv(timeout=0.5))
+                    self._handle_message(message, ws)
+                except TimeoutError:
+                    pass
+
+    def _handle_message(self, message, ws):
+        if message["type"] == "conversation_initiation_metadata":
+            event = message["conversation_initiation_metadata_event"]
+            assert self._conversation_id is None
+            self._conversation_id = event["conversation_id"]
+        elif message["type"] == "audio":
+            event = message["audio_event"]
+            if int(event["event_id"]) <= self._last_interrupt_id:
+                return
+            audio = base64.b64decode(event["audio_base_64"])
+            self.audio_interface.output(audio)
+        elif message["type"] == "agent_response":
+            if self.callback_agent_response:
+                event = message["agent_response_event"]
+                self.callback_agent_response(event["agent_response"].strip())
+        elif message["type"] == "agent_response_correction":
+            if self.callback_agent_response_correction:
+                event = message["agent_response_correction_event"]
+                self.callback_agent_response_correction(event["original_agent_response"].strip(), event["corrected_agent_response"].strip())
```

Comment:
Is this indented correctly? Github shows weirdly

Reply:
It's fine, just long line. Ran `ruff format` which split it up a bit.